src/annotation/decorator_utils.ts

```diff
@@ -1,67 +1,99 @@
+}
+
+function _ensureNoMetadataKeyDuplicates(metadata: interfaces.MetadataOrMetadataArray):interfaces.Metadata[]{
+    let metadatas: interfaces.Metadata[] = [];
+    if(Array.isArray(metadata)){
+        metadatas = metadata;
+        const duplicate = getFirstArrayDuplicate(metadatas.map(md => md.key));
+        if(duplicate !== undefined) {
+            throw new Error(`${ERROR_MSGS.DUPLICATED_METADATA} ${duplicate.toString()}`);
+        }
+    }else{
+        metadatas = [metadata];
+    }
+    return metadatas;
 }
 
 function _tagParameterOrProperty(
     metadataKey: string,
     annotationTarget: any,
-    propertyName: string,
-    metadata: interfaces.Metadata,
-    parameterIndex?: number
+    key: string | symbol,
+    metadata: interfaces.MetadataOrMetadataArray,
 ) {
+    const metadatas: interfaces.Metadata[] = _ensureNoMetadataKeyDuplicates(metadata);
 
-    let paramsOrPropertiesMetadata: interfaces.ReflectResult = {};
-    const isParameterDecorator = (typeof parameterIndex === "number");
-    const key: string = (parameterIndex !== undefined && isParameterDecorator) ? parameterIndex.toString() : propertyName;
-
-    // if the decorator is used as a parameter decorator, the property name must be provided
-    if (isParameterDecorator && propertyName !== undefined) {
-        throw new Error(ERROR_MSGS.INVALID_DECORATOR_OPERATION);
-    }
-
+    let paramsOrPropertiesMetadata:any = {};
     // read metadata if available
     if (Reflect.hasOwnMetadata(metadataKey, annotationTarget)) {
         paramsOrPropertiesMetadata = Reflect.getMetadata(metadataKey, annotationTarget);
     }
 
-    // get metadata for the decorated parameter by its index
-    let paramOrPropertyMetadata: interfaces.Metadata[] = paramsOrPropertiesMetadata[key];
+    let paramOrPropertyMetadata: interfaces.Metadata[] | undefined = paramsOrPropertiesMetadata[key];
 
-    if (!Array.isArray(paramOrPropertyMetadata)) {
+    if (paramOrPropertyMetadata === undefined) {
         paramOrPropertyMetadata = [];
     } else {
         for (const m of paramOrPropertyMetadata) {
-            if (m.key === metadata.key) {
+            if (metadatas.some(md => md.key === m.key)) {
                 throw new Error(`${ERROR_MSGS.DUPLICATED_METADATA} ${m.key.toString()}`);
             }
         }
     }
 
     // set metadata
-    paramOrPropertyMetadata.push(metadata);
+    paramOrPropertyMetadata.push(...metadatas);
     paramsOrPropertiesMetadata[key] = paramOrPropertyMetadata;
     Reflect.defineMetadata(metadataKey, paramsOrPropertiesMetadata, annotationTarget);
 
 }
+type ParameterOrPropertyDecorator = (
+    target: any,
+    targetKey: string | symbol,
+    indexOrPropertyDescriptor?: number | TypedPropertyDescriptor<unknown>
+) => void;
+function createTaggedDecoratorInternal(
+    metadata:interfaces.MetadataOrMetadataArray,
+    callback?:ParameterOrPropertyDecorator
```

Comment:
This callback is used only on the `@inject` decorator, we should remove this parameter in favor of creating an arrow in the `@inject` decorator:

```ts
function inject(serviceIdentifier: ServiceIdentifierOrFunc) {
  return (target, targetKey, indexOrPropertyDescriptor) => {
      if (serviceIdentifier === undefined) {
          throw new Error(UNDEFINED_INJECT_ANNOTATION(target.name));
      }

      return createTaggedDecoratorInternal(new Metadata(METADATA_KEY.INJECT_TAG, serviceIdentifier)(target, targetKey, indexOrPropertyDescriptor);
  };
}
```

Reply:
Agreed

Reply:
I think that multiinject should behave the same and throw for missing serviceidentifierorfunc and that it should go in this pr.

Reply:
Sure :)